cassandra-diagnostics-core/src/main/java/io/smartcat/cassandra/diagnostics/module/hiccup/HiccupModule.java

```diff
@@ -0,0 +1,187 @@
+
+    private static final String DEFAULT_MEASUREMENT_NAME = "hiccup";
+
+    private static final String HICCUP_THREAD_NAME = "hiccup-thread";
+
+    private static final double NANOS_IN_MILLIS = 1000000d;
+
+    private final HiccupConfiguration config;
+
+    private final String service;
+
+    private final Thread thread;
+
+    /**
+     * Constructor.
+     *
+     * @param configuration       Module configuration
+     * @param reporters           Reporter list
+     * @param globalConfiguration Global diagnostics configuration
+     * @throws ConfigurationException in case the provided module configuration is not valid
+     */
+    public HiccupModule(ModuleConfiguration configuration, List<Reporter> reporters,
+            GlobalConfiguration globalConfiguration) throws ConfigurationException {
+        super(configuration, reporters, globalConfiguration);
+
+        this.config = HiccupConfiguration.create(configuration.options);
+        this.service = configuration.getMeasurementOrDefault(DEFAULT_MEASUREMENT_NAME);
+
+        logger.info("Hiccup module initialized with {} {} reporting period.", config.period(),
+                config.timeunit().name());
+
+        this.thread = new Thread(new HiccupMeter(), HICCUP_THREAD_NAME);
+        this.thread.start();
+    }
+
+    @Override
+    public void stop() {
+        logger.trace("Stopping hiccup module.");
+        try {
+            thread.join(1000);
+        } catch (InterruptedException e) {
+            logger.error("Hiccup thread failed to stop", e);
+        }
+    }
+
+    /**
+     * Hiccup meter running hiccup recorder and collecting measurements.
+     */
+    private class HiccupMeter implements Runnable {
+
+        private HiccupRecorder hiccupRecorder = null;
+        private Histogram intervalHistogram = null;
+
+        @Override
+        public void run() {
+            final SingleWriterRecorder recorder = new SingleWriterRecorder(config.lowestTrackableValue(),
+                    config.highestTrackableValue(), config.numberOfSignificantValueDigits());
+
+            final long uptimeAtInitialStartTime = ManagementFactory.getRuntimeMXBean().getUptime();
+            long now = System.currentTimeMillis();
+            long jvmStartTime = now - uptimeAtInitialStartTime;
+
+            hiccupRecorder = new HiccupRecorder(config, recorder);
+
+            try {
+                final long startTime;
+
+                if (config.startDelayInMs() > 0) {
+                    // Run hiccup recorder during startDelayInMs time to let code warm up:
+                    hiccupRecorder.start();
+                    while (config.startDelayInMs() > System.currentTimeMillis() - jvmStartTime) {
+                        Thread.sleep(100);
+                    }
+                    hiccupRecorder.terminate();
+                    hiccupRecorder.join();
+
+                    recorder.reset();
+                    hiccupRecorder = new HiccupRecorder(config, recorder);
```

Comment:
I would extract this to a `warmup` method since level of implementation is not needed in run method.

```
if (config.startDelayInMs() > 0) {
   warmup()
}
```